addon/tracked.ts

```diff
@@ -1,143 +1,180 @@
-
-  let values = new WeakMap();
-
-  let get;
-  if (typeof initializer === 'function') {
-    get = function(this: object) {
-      if (values.has(this)) {
-        return values.get(this);
-      } else {
-        let value = initializer.call(this);
-        values.set(this, value);
-        return value;
-      }
-    };
-  } else {
-    get = function(this: object) {
-      return values.get(this);
-    }
+function getDependentsOf<O extends object>([obj, key]: Field<O>): [
+  any,
+  ReadonlySet<Key>
+][] {
+  if (DEPENDENCIES.has(obj) && DEPENDENCIES.get(obj).has(key)) {
+    return Array.from(DEPENDENCIES.get(obj).get(key));
   }
 
-  return {
-    configurable: true,
-    // TODO: correcting a misspelling caused chrome to error
-    // writable: true,
-
-    get,
-    set(value) {
-      if (typeof key === 'string') {
-        values.set(this, value);
-        Ember.notifyPropertyChange(this, key);
-      } else if (DEBUG) {
-        throw new Error(`@tracked - unsupported property type ${String(key)}`);
-      }
+  return [];
+}
+
+function invalidateDependentsOf<O extends object>(field: Field<O>) {
+  for (const [obj, keys] of getDependentsOf(field)) {
+    for (const key of keys) {
+      notifyPropertyChange(obj, key);
     }
-  };
+  }
 }
 
+function clearDependencies([obj, key]: Field<any>) {
+  // if (DEPENDENCIES.has(obj)) {
+  //   DEPENDENCIES.get(obj).delete(key);
+  // }
+}
 
-function _tracked(target: object, key: string | symbol, descriptor?: PropertyDescriptor, dependencies?: string[]): PropertyDescriptor {
-  // descriptor is undefined for typescript class fields
-  if (descriptor === undefined || 'initializer' in descriptor) {
-    return installTrackedProperty(target, key, descriptor);
-  } else {
-    return descriptorForTrackedComputedProperty(target, key, descriptor, dependencies);
+export const tracked = decoratorWithParams((desc, params: string[] = []) => {
+  assert(
+    '@tracked: You can only apply this decorator to class fields.',
+    desc.kind === 'field' || (desc.kind === 'method' && !desc.value)
+  );
+
+  const { descriptor, key, initializer } = desc;
+  const { get, set } = descriptor;
+
+  delete desc.initializer;
+  desc.kind = 'method';
+  delete descriptor.writable;
+
+  if (set) {
+    assert(
+      `@tracked: Marking property '${key} as tracked does not make sense, since it only has a setter.`,
+      get
+    );
+    descriptor.set = trackSet(key, set);
+  }
+  if (get) {
+    descriptor.get = trackGet(key, get);
+    if (DEBUG && !set) {
+      descriptor.set = function(value: any) {
+        throw new TypeError(
+          `@tracked: You cannot set the tracked property '${key}' to '${value}', because it only has a getter. Add a setter in order to set the value.`
+        );
+      };
+    }
+  }
+  if (!get && !set) {
+    descriptor.set = trackSet(key);
+    descriptor.get = trackGet(key);
+  }
+
+  if (initializer) {
+    desc.extras = [
+      {
+        ...INITIALIZER,
```

Comment:
`kind: 'initializer'` is not currently supported in Babel and is pending finalization of the spec: https://github.com/tc39/proposal-decorators/issues/182